File: 0091_decode_ways.py
# https://leetcode-cn.com/problems/decode-ways/

class Solution:
    def numDecodings(self, s):
        """
        Inputs:
            s: str
        Outputs:
            res: int
        """
        
        if s[0] == '0':
            return 0
        
        n = len(s)
        dp = [1 for _ in range(n+1)]

        for i in range(1,n):
            # 4 conditions.
            tmp = int(s[i-1:i+1])
            if s[i] == '0':
                if tmp > 20 ot tmp == 0:
                    return 0
                else:
                    dp[i] = dp[i-2]
            else:
                if tmp > 26 or tmp < 10:
                    dp[i] = dp[i-1]
                else:
                    dp[i] = dp[i-1] + dp[i-2]
        return dp[-2]

File: Cpp/0091_decode_ways.cpp
// https://leetcode.cn/problems/decode-ways/

class Solution {
public:
    int numDecodings(string s) {
        if (s[0] == '0') {
            return 0;
        }
        
        int res = 1, res_last = 1, res_last_last = 1;
        for (int i = 1; i < s.size(); ++i) {
            int num1 = s[i-1] - '0', num2 = s[i] - '0';
            int val = 10 * num1 + num2;
            
            if (num2 == 0) {
                if ((val == 0) || (val >= 30)) {
                    return 0;
                }
                else {
                    res = res_last_last;
                }
            }
            else {
                if ((val >= 11) && (val <= 26)) {
                    res = res_last_last + res_last;
                }
                else {
                    res = res_last;
                }
            }
            
            res_last_last = res_last;
            res_last = res;
        }
        
        return res;
    }
};
